kong/cluster_events/init.lua

```diff
@@ -219,67 +277,27 @@ local function poll(self)
 
   min_at = min_at - self.poll_offset - 0.001
 
-  local max_at = ngx_now()
+  log(DEBUG, "polling events from: ", min_at)
 
-  log(DEBUG, "polling events from: ", min_at, " to: ", max_at)
-
-  for rows, err, page in self.strategy:select_interval(self.channels, min_at, max_at) do
+  for rows, err, page in self.strategy:select_interval(self.channels, min_at) do
     if err then
       return nil, "failed to retrieve events from DB: " .. err
     end
 
+    local count = #rows
+
     if page == 1 then
-      local ok, err = self.shm:safe_set(CURRENT_AT_KEY, max_at)
+      local at = count > 0 and rows[1].now or self.strategy:server_time() or ngx_now()
```

Comment:
I don't think we should ever get `self.strategy:server_time()` here. Assuming that we used it, this timestamp is potentially greater than that of the last event returned, meaning that any events in the time window between `select_interval` and `server_time` would not be processed. That time window is usually tiny, but DB operations may yield, so an arbitrary amount of time can pass between lines 282 and 290.

I think `CURRENT_AT_KEY` should be set to the largest `row.at` received by this loop (or rather, the latest one executed before a failure, if we care about `process_event` error codes and want to retry). Since `poll` (generally!) runs locked and also does the `ran` checks, it shouldn't make a difference if it is set at the end, or the beginning of this loop.

If no rows were returned, `CURRENT_AT_KEY` can stay unchanged. It will eventually move forward when the next event arrives.

Reply:
@hishamhm I think I asked about this here:
https://github.com/Kong/kong/pull/5247#pullrequestreview-321482345

This didn't change. I can change it though.

Reply:
Yes, this changed so that we only advance the time when there has been events.